File: ios/AppVideosManager.swift
//
//  AppVideosManager.swift
//  Video
//
//  Created by sergeymild on 11/08/2023.
//  Copyright © 2023 Facebook. All rights reserved.
//

import Foundation

public typealias LookyVideoView = VideoViewSwift

public class VideoChannel {
    var videos: [String: LookyVideoView] = [:]
    var laterRestore: String?
    var backgroundRestore: String?

    var currentPlaying: (key: String, value: LookyVideoView)? {
        for v in videos {
            if v.value._paused { continue }
            return v
        }
        return nil
    }

    var forRestore: LookyVideoView? {
        if let backgroundRestore {
            return videos[backgroundRestore]
        }
        return nil
    }

    func pauseAllVideos() {
        for video in videos {
            if video.value._paused { continue }
            video.value.setPaused(true)
        }
    }

    func toggleMuted(_ muted: Bool) {
        for video in videos {
            video.value.updateMuted(muted)
        }
    }

    func video(for videoId: String) -> LookyVideoView? {
        return videos[videoId]
    }
}

@objc
public class AppVideosManager: NSObject {
    @objc
    public static let shared = AppVideosManager()
    public var channels: [String: VideoChannel] = [:]

    public func addVideo(_ video: LookyVideoView, nativeID: String) {
        let parts = nativeID.components(separatedBy: ":")
        let channelName = parts[0]
        let id = parts[1]

        let channel = getChannel(name: channelName)
        channel?.videos[id] = video

        //debugPrint("🍓 addVideo", channelName, id)

        if channel?.laterRestore == id {
            video.setPaused(false)
            channel?.laterRestore = nil
        } else {
            video.showPoster(show: true)
        }
    }

    public func removeVideo(_ nativeID: String) {
        let parts = nativeID.components(separatedBy: ":")
        let channelName = parts[0]
        let id = parts[1]
        //debugPrint("🍓 removeVideo", channelName, id)

        let channel = getChannel(name: channelName)
        channel?.videos.removeValue(forKey: id)
        if channel?.videos.isEmpty == true {
            channels.removeValue(forKey: channelName)
        }
    }
}


// MARK: public api
extension AppVideosManager {
    @objc
    public func playVideo(_ channelName: String, videoId: String) {
        guard let channel = getChannel(name: channelName) else { return }
        if channel.currentPlaying?.key == videoId { return }
        self.pauseCurrentPlaying()
        if let video = channel.video(for: videoId), video._paused {
            debugPrint("🍓 playVideo", channelName, videoId)
            video.setPaused(false)
        }
    }

    @objc
    public func pauseVideo(_ channelName: String, videoId: String) {
        guard let channel = getChannel(name: channelName) else { return }
        if let video = channel.video(for: videoId), !video._paused {
            debugPrint("🍓 pauseVideo", channelName, videoId)
            video.setPaused(true)
        }
    }

    @objc
    public func togglePlayInBackground(_ channelName: String?, playInBackground: Bool) {
        guard let videoChannel = getChannel(name: channelName) else { return }
        if playInBackground {
            if let video = findFirstPlayingVideo(channelName: channelName) {
                debugPrint("🍓 togglePlayInBackground", channelName as Any, videoId(video) as Any)
                video.applicationDidEnterBackground()
                videoChannel.backgroundRestore = videoId(video)
            }
        } else {
            videoChannel.forRestore?.applicationWillEnterForeground()
            videoChannel.backgroundRestore = nil
        }

    }

    @objc
    public func restoreLastPlaying(_ channelName: String?, shouldSeekToStart: Bool) {
        pauseAllVideos()
        debugPrint("🍓 restoreLastPlaying.knownChannel", channelName as Any)
        if let channelName {
            let videoChannel = getChannel(name: channelName)
            if let restore = videoChannel?.laterRestore {
                togglePlay(channel: channelName, videoId: restore, seekToStart: true)
            }
            return
        }

        for entry in channels {
            if entry.value.laterRestore == nil { continue }
            togglePlay(
                channel: entry.key,
                videoId: entry.value.laterRestore!,
                seekToStart: true
            )
            entry.value.laterRestore = nil
        }
    }

    @objc
    public func pauseCurrentPlayingWithLaterRestore(_ channelName: String?) {
        if let channelName {
            guard let channel = getChannel(name: channelName) else { return }
            if let video = channel.currentPlaying?.value {
                channel.laterRestore = videoId(video)
                video.setPaused(true)
                debugPrint("🍓 pauseCurrentPlayingWithLaterRestore.knownChannel", channelName as Any, videoId(video) as Any)
            }
            return;
        }

        for entry in channels {
            let channel = entry.value
            let video = channel.currentPlaying?.value
            channel.laterRestore = videoId(video)
            video?.setPaused(true)
            debugPrint("🍓 pauseCurrentPlayingWithLaterRestore", entry.key, videoId(video) as Any)
        }
    }

    @objc
    public func togglePlayVideo(_ channelName: String, videoId: String) {
        let video = findFirstPlayingVideo(channelName: channelName)
        if video != nil { pauseCurrentPlaying() }
        else { playVideo(channelName, videoId: videoId) }
    }

    @objc
    public func toggleVideosMuted(_ muted: Bool) {
        DispatchQueue.main.async {
            for entry in self.channels {
                entry.value.toggleMuted(muted)
            }
        }
    }

    @objc
    public func pauseCurrentPlaying() {
        for entry in channels {
            let channel = entry.value
            if let video = channel.currentPlaying?.value {
                debugPrint("🍓 pauseCurrentPlaying", video.nativeID ?? "")
                video.setPaused(true)
            }
        }
    }

    @objc
    public func pauseAll(_ channelName: String) {
        guard let channel = getChannel(name: channelName) else {
            return
        }

        for entry in channel.videos {
            entry.value.setPaused(true)
        }
    }

    @objc
    public func playAll(_ channelName: String) {
        guard let channel = getChannel(name: channelName) else {
            return
        }

        for entry in channel.videos {
            entry.value.setPaused(false)
        }
    }

    @objc
    public func isPaused(_ channelName: String, videoId: String) -> Bool {
        return getChannel(name: channelName)?.video(for: videoId)?._paused == true
    }

    @objc
    public func isMuted(_ channelName: String, videoId: String) -> Bool {
        return getChannel(name: channelName)?.video(for: videoId)?._muted == true
    }

    @objc
    public func seek(_ channelName: String, videoId: String, duration: Double) {
        getChannel(name: channelName)?.video(for: videoId)?.seekTo(duration: duration)
    }
}


// MARK: private api
extension AppVideosManager {
    private func togglePlay(channel: String, videoId: String, seekToStart: Bool) {
        guard let videoChannel = getChannel(name: channel) else {
            return
        }

        let playingVideo = findFirstPlayingVideo(channelName: channel);
        let video = videoChannel.video(for: videoId)
        let playingVideId = self.videoId(playingVideo)

        // pause current playing video
        if let playingVideId, playingVideId != videoId {
            playingVideo?.setPaused(true)
        }

        if let video {
            if seekToStart { video.seekToStart() }
            video.setPaused(false)
        }
    }

    private func videoId(_ video: LookyVideoView?) -> String? {
        guard let video = video else { return nil }
        let parts = video.nativeID.components(separatedBy: ":")
        return parts[1]
    }

    private func getChannel(name: String?) -> VideoChannel? {
        guard let name else { return nil }
        let channel = channels[name] ?? VideoChannel()
        channels[name] = channel
        return channel
    }

    private func pauseVideo(channel: VideoChannel) {
        channel.currentPlaying?.value.setPaused(true)
    }

    private func pauseAllVideos() {
        for entry in channels {
            entry.value.pauseAllVideos()
        }
    }

    private func findFirstPlayingVideo() -> LookyVideoView? {
        for entry in channels {
            if let v = entry.value.currentPlaying { return v.value }
        }
        return nil
    }

    private func findFirstPlayingVideo(channelName: String?) -> LookyVideoView? {
        guard let name = channelName else { return nil }
        guard let channel = getChannel(name: name) else { return nil }
        return channel.currentPlaying?.value
    }
}

File: android/src/main/java/com/video/AppVideosManager.kt
package com.video

import android.os.Handler
import android.os.Looper
import java.util.concurrent.ConcurrentHashMap

typealias LookyVideoView = VideoView

class VideoChannel {
  var videos = ConcurrentHashMap<String, LookyVideoView>()
  var laterRestore: String? = null
  var backgroundRestore: String? = null

  val currentPlaying: MutableMap.MutableEntry<String, LookyVideoView>?
    get() {
      for (video in videos) {
        if (video.value.isSafePlaying()) return video
      }
      return null
    }

  val forRestore: LookyVideoView?
    get() {
      backgroundRestore ?: return null
      return videos[backgroundRestore]
    }

  fun pauseAllVideos() {
    for (entry in videos) {
      if (entry.value.isSafePlaying())
        entry.value.setPausedModifier(true)
    }
  }

  fun toggleMuted(muted: Boolean) {
    for (entry in videos) {
      entry.value.setMutedModifier(muted)
    }
  }

  fun video(videoId: String) = videos[videoId]
}

class AppVideosManager {
  companion object {
    val shared = AppVideosManager()
  }

  val channels = ConcurrentHashMap<String, VideoChannel>()
  val handler = Handler(Looper.getMainLooper())

  fun addVideo(video: LookyVideoView, nativeID: String) {
    val parts = nativeID.split(":")
    val channelName = parts[0]
    val id = parts[1]
    println("🍓 addVideo $channelName $id")

    val channel = getChannel(channelName)?.also {
      it.videos[id] = video
    }

    if (channel?.laterRestore == id) {
      video.setPausedModifier(false)
      channel.laterRestore = null
    } else {
      video.showPoster(true)
    }
  }

  fun removeVideo(nativeID: String) {
    val parts = nativeID.split(":")
    val channelName = parts[0]
    val id = parts[1]
    println("🍓 removeVideo $channelName $id")

    val channel = getChannel(channelName)
    channel?.videos?.remove(id)
    if (channel?.videos?.isEmpty() == true) {
      channels.remove(channelName)
    }
  }
}

// MARK: public api
fun AppVideosManager.playVideo(channelName: String, videoId: String) {
  val channel = getChannel(channelName) ?: return
  if (channel.currentPlaying?.key == videoId) { return }
  pauseCurrentPlaying()
  val video = channel.video(videoId) ?: return
  println("🍓 playVideo $channelName, $videoId")
  video.setPausedModifier(false)
}

fun AppVideosManager.pauseVideo(channelName: String, videoId: String) {
  val channel = getChannel(channelName) ?: return
  val video = channel.video(videoId) ?: return
  if (!video.isSafePlaying()) return
  println("🍓 pauseVideo $channelName, $videoId")
  video.setPausedModifier(true)
}

fun AppVideosManager.togglePlayInBackground(channelName: String?, playInBackground: Boolean) {}

fun AppVideosManager.restoreLastPlaying(channelName: String?, shouldSeekToStart: Boolean) {
  pauseAllVideos()
  println("🍓 restoreLastPlaying $channelName")
  if (channelName != null) {
    val videoChannel = getChannel(channelName)
    val restore = videoChannel?.laterRestore
    if (restore != null) {
      togglePlay(channel = channelName, videoId = restore, seekToStart = true)
    }
    return
  }

  for (entry in channels) {
    if (entry.value.laterRestore == null) continue
    togglePlay(
      channel = entry.key,
      videoId = entry.value.laterRestore!!,
      seekToStart = true
    )
    entry.value.laterRestore = null
  }
}

fun AppVideosManager.pauseCurrentPlayingWithLaterRestore(channelName: String?) {
  if (channelName != null) {
    val videoChannel = getChannel(channelName) ?: return
    val video = findFirstPlayingVideo(channelName) ?: return
    videoChannel.laterRestore = videoId(video)
    video.setPausedModifier(true)
    return
  }

  for (entry in channels) {
    val channel = entry.value
    val video = findFirstPlayingVideo(entry.key)
    channel.laterRestore = videoId(video)
    video?.setPausedModifier(true)
  }
}

fun AppVideosManager.togglePlayVideo(channelName: String, videoId: String) {
  val video = findFirstPlayingVideo(channelName)
  if (video != null) { pauseCurrentPlaying() }
  else { playVideo(channelName, videoId) }
}

fun AppVideosManager.toggleVideosMuted(muted: Boolean) {
  for (entry in channels) {
    entry.value.toggleMuted(muted)
  }
}

fun AppVideosManager.pauseCurrentPlaying() {
  pauseAllVideos()
}

fun AppVideosManager.isPaused(channel: String, videoId: String): Boolean {
  return getChannel(channel)?.video(videoId)?.mPaused == true
}

fun AppVideosManager.isMuted(channel: String, videoId: String): Boolean {
  return getChannel(channel)?.video(videoId)?.mMuted == true
}

fun AppVideosManager.seek(channel: String, videoId: String, duration: Double) {
  getChannel(channel)?.video(videoId)?.seekTo((duration / 1000.0).toInt())
}

fun AppVideosManager.playAll(channelName: String) {
  val channel = getChannel(channelName) ?: return
  for (video in channel.videos) {
    video.value.setPausedModifier(false)
  }
}

fun AppVideosManager.pauseAll(channelName: String) {
  val channel = getChannel(channelName) ?: return
  for (video in channel.videos) {
    video.value.setPausedModifier(true)
  }
}


fun AppVideosManager.togglePlay(channel: String, videoId: String, seekToStart: Boolean) {
  val videoChannel = getChannel(channel) ?: return

  val playingVideo = findFirstPlayingVideo(channel)
  val video = videoChannel.video(videoId)
  val playingVideId = videoId(playingVideo)

  // pause current playing video
  if (playingVideId != null && playingVideId != videoId) {
    playingVideo?.setPausedModifier(true)
  }

  if (video != null) {
    if (seekToStart) video.seekTo(0)
    video.setPausedModifier(false)
  }
}

fun AppVideosManager.videoId(video: LookyVideoView?): String? {
  video ?: return null
  val parts = (video.getTag(R.id.view_tag_native_id) as? String)?.split(":")
  return parts?.get(1)
}

fun AppVideosManager.getChannel(name: String?): VideoChannel? {
  name ?: return null
  val channel = channels[name] ?: VideoChannel()
  channels[name] = channel
  return channel
}

fun AppVideosManager.pauseVideo(channel: VideoChannel) {
  channel.currentPlaying?.value?.setPausedModifier(true)
}

fun AppVideosManager.pauseAllVideos() {
  for (entry in channels) {
    entry.value.pauseAllVideos()
  }
}

fun AppVideosManager.findFirstPlayingVideo(): LookyVideoView? {
  for (entry in channels) {
    entry.value.currentPlaying?.let { return it.value }
  }
  return null
}

fun AppVideosManager.findFirstPlayingVideo(channelName: String?): LookyVideoView? {
  channelName ?: return null
  val channel = getChannel(channelName) ?: return null
  return channel.currentPlaying?.value
}
